File: kotlin-electron/karakum/src/jsMain/resources/plugins/convertNodeJsQualifiedName.js
import ts from "typescript";

function convertNodePackageQualifiedName(node, context, render) {
    if (
        ts.isIdentifier(node)
        && node.text === "EventEmitter"
    ) {
        return `node.events.EventEmitter`
    }

    return null
}

export default function (node, context, render) {
    if (
        ts.isQualifiedName(node)
        && ts.isIdentifier(node.left)
        && node.left.text === "NodeJS"
    ) {
        const result = convertNodePackageQualifiedName(node.right, context, render)
        if (result !== null) return result

        return `node.${render(node.right)}`
    }

    if (
        ts.isPropertyAccessExpression(node)
        && ts.isIdentifier(node.expression)
        && node.expression.text === "NodeJS"
    ) {
        const result = convertNodePackageQualifiedName(node.name, context, render)
        if (result !== null) return result

        return `node.${render(node.name)}`
    }

    return null
}

File: kotlin-electron/karakum/src/jsMain/kotlin/electron/karakum/plugins/convertNodeJsQualifiedName.kt
package electron.karakum.plugins

import electron.karakum.util.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import typescript.isIdentifier
import typescript.isPropertyAccessExpression
import typescript.isQualifiedName

val convertNodeJsQualifiedName = createPlugin { node, _, render ->
    nullable {
        ensure(isQualifiedName(node))

        val left = node.left
        ensure(isIdentifier(left))
        ensure(left.text == "NodeJS")

        when (node.right.text) {
            "EventEmitter" -> "node.events.EventEmitter"
            else -> "node.${render(node.right)}"
        }
    } ?: nullable {
        ensure(isPropertyAccessExpression(node))

        val expression =  node.expression
        ensure(isIdentifier(expression))
        ensure(expression.text == "NodeJS")

        val name =  node.name
        ensure(isIdentifier(name))

        when (name.text) {
            "EventEmitter" -> "node.events.EventEmitter"
            else -> "node.${render(name)}"
        }
    }
}
